### PR TITLE
Content hash based reordering

### PR DESCRIPTION
**Pull request by keyboardDrummer**
Reorders top level axioms based on a hash of their content. This enables reordering Boogie members without effecting solver input.

Based on #452

### CODE REVIEW

File: Source/Core/SMTLibOptions.cs
Comment (RustanLeino) [inline Source/Core/SMTLibOptions.cs:25]: "all identifiers" refers to both global and local names, right?

File: Source/Core/SMTLibOptions.cs
Comment (RustanLeino) [inline Source/Core/SMTLibOptions.cs:29]: Is there any reason _not_ to always do this transformation? It seems good for everyone, for perhaps this should just always be on. (Or, if this causes the output to use the generated names, then I'm doubtful that anyone could afford to use it.)

**Review by RustanLeino — DISMISSED** [review]
I love this feature.

Well, I _think_ I love it. I'm basing that love on the believe that this does not change the output. Is that right? For example, with this feature engaged, will the Dafny counterexample display in the IDE continue working without change?

Also, I'm clicking "Approve", but I expect the changes to this PR's dependent PRs to be addressed first.

File: Source/Core/SMTLibOptions.cs
Comment (keyboardDrummer) [reply Source/Core/SMTLibOptions.cs:25]: Jep

**Review by RustanLeino — DISMISSED** [review]

**Comment by shazqadeer** [comment]
In PR #452, the goal of this PR is described as:

> Reorder top level declarations based on a content hash.

However, the description of this PR says:

> Reorders top level axioms based on a hash of their content.

Are you reordering only axioms or all declarations?  Please make the descriptions consistent.

**Comment by keyboardDrummer** [comment]
> In PR #452, the goal of this PR is described as:
> 
> > Reorder top level declarations based on a content hash.
> 
> However, the description of this PR says:
> 
> > Reorders top level axioms based on a hash of their content.
> 
> Are you reordering only axioms or all declarations? Please make the descriptions consistent.

I'm only reordering the axioms based on a content hash. Reordering the axioms turns out to be enough since the order in which top level functions and constants are translated to SMT is not based on their declaration order, but by the order in which they're discovered when traversing the axioms.

**Review by shazqadeer — APPROVED** [review]

**Review by robin-aws — APPROVED** [review]
Re-approving after previous approvals were dismissed by resolving merge conflicts.

**Review by shazqadeer — APPROVED** [review]

**Pull request merged**
